x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecycleExecutionState.java

```diff
@@ -0,0 +1,253 @@
+                                    String stepInfo, String phaseDefinition, long indexCreationDate,
+                                    long phaseTime, long actionTime, long stepTime) {
+        this.phase = phase;
+        this.action = action;
+        this.step = step;
+        this.failedStep = failedStep;
+        this.stepInfo = stepInfo;
+        this.phaseDefinition = phaseDefinition;
+        this.indexCreationDate = indexCreationDate;
+        this.phaseTime = phaseTime;
+        this.actionTime = actionTime;
+        this.stepTime = stepTime;
+    }
+
+    /**
+     * Retrieves the execution state from an {@link IndexMetaData} based on the
+     * custom metadata.
+     * @param indexMetaData The metadata of the index to retrieve the execution
+     *                      state from.
+     * @return The execution state of that index.
+     */
+    public static LifecycleExecutionState fromIndexMetadata(IndexMetaData indexMetaData) {
+        Map<String, String> customData = indexMetaData.getCustomData(ILM_CUSTOM_METADATA_KEY);
+        customData = customData == null ? new HashMap<>() : customData;
+        return fromCustomMetadata(customData);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static Builder builder(LifecycleExecutionState state) {
+        return new Builder()
+            .setPhase(state.phase)
+            .setAction(state.action)
+            .setStep(state.step)
+            .setFailedStep(state.failedStep)
+            .setStepInfo(state.stepInfo)
+            .setPhaseDefinition(state.phaseDefinition)
+            .setIndexCreationDate(state.indexCreationDate)
+            .setPhaseTime(state.phaseTime)
+            .setActionTime(state.actionTime)
+            .setStepTime(state.stepTime);
+    }
+
+    static LifecycleExecutionState fromCustomMetadata(Map<String, String> customData) {
+        return new Builder()
+            .setPhase(customData.getOrDefault(PHASE, ""))
+            .setAction(customData.getOrDefault(ACTION, ""))
+            .setStep(customData.getOrDefault(STEP, ""))
+            .setFailedStep(customData.getOrDefault(FAILED_STEP, ""))
+            .setStepInfo(customData.getOrDefault(STEP_INFO, ""))
+            .setPhaseDefinition(customData.getOrDefault(PHASE_DEFINITION, ""))
+            .setIndexCreationDate(Long.parseLong(customData.getOrDefault(INDEX_CREATION_DATE, "-1")))
+            .setPhaseTime(Long.parseLong(customData.getOrDefault(PHASE_TIME, "-1")))
+            .setActionTime(Long.parseLong(customData.getOrDefault(ACTION_TIME, "-1")))
+            .setStepTime(Long.parseLong(customData.getOrDefault(STEP_TIME, "-1")))
+            .build();
+    }
+
+    /**
+     * Converts this object to an immutable map representation for use with
+     * {@link IndexMetaData.Builder#putCustom(String, Map)}.
+     * @return An immutable Map representation of this execution state.
+     */
+    public Map<String, String> asMap() {
```

Comment:
I'm not sure how I feel about these, in particular, I think `imd` and `imdNew` should be exactly the same metadata:

```java
IndexMetaData imd = state.metaData().index(i).getCustomData("ilm");
LifecycleExecutionState les = LifecycleExecutionState.fromIndexMetaData(imd);
IndexMetaData imdNew = IndexMetaData.builder(imd).putCustom("ilm", les.asMap()).build);
```

But right now we are substituting not-found values like `""` and `"-1"` in the event that there are missing keys, I think we should only store keys that have been set

Reply:
My personal preference would be to return `null` so that

```java
LifecycleExecutionState.builder()
   .setPhase(oldState.getPhase())
   .build();
```

Returns a state with the phase set to `null` and elides it from the `asMap` output. Otherwise we run into the same problem when a builder is used

Reply:
Oh, good point - I'll do that then.